Tests/NIOPosixTests/NIOScheduledCallbackTests.swift

```diff
@@ -0,0 +1,151 @@
+        let handler = MockTimerHandler()
+        let handle = loop.scheduleCallback(in: .milliseconds(42), handler: handler)
+
+        handle.cancel()
+        XCTAssertEqual(handler.firedCount, 0)
+
+        loop.advanceTime(by: .milliseconds(43))
+        XCTAssertEqual(handler.firedCount, 0)
+    }
+}
+
+final class NIOAsyncTestingEventLoopScheduledCallbackTests: XCTestCase {
+
+    func testFired() async {
+        let loop = NIOAsyncTestingEventLoop()
+
+        let handler = MockTimerHandler()
+        XCTAssertEqual(handler.firedCount, 0)
+
+        _ = loop.scheduleCallback(in: .milliseconds(42), handler: handler)
+        XCTAssertEqual(handler.firedCount, 0)
+
+        await loop.advanceTime(by: .milliseconds(41))
+        XCTAssertEqual(handler.firedCount, 0)
+
+        await loop.advanceTime(by: .milliseconds(1))
+        XCTAssertEqual(handler.firedCount, 1)
+
+        await loop.advanceTime(by: .milliseconds(1))
+        XCTAssertEqual(handler.firedCount, 1)
+    }
+
+    func testCancelled() async {
+        let loop = NIOAsyncTestingEventLoop()
+        let handler = MockTimerHandler()
+        let handle = loop.scheduleCallback(in: .milliseconds(42), handler: handler)
+
+        handle.cancel()
+        XCTAssertEqual(handler.firedCount, 0)
+
+        await loop.advanceTime(by: .milliseconds(43))
+        XCTAssertEqual(handler.firedCount, 0)
+    }
+}
+
+final class MTELGScheduledCallbackTests: XCTestCase {
+
+    func testFired() async throws {
+        let loop = MultiThreadedEventLoopGroup.singleton.next()
+
+        let handler = MockTimerHandler()
+        XCTAssertEqual(handler.firedCount, 0)
+
+        _ = loop.scheduleCallback(in: .milliseconds(1), handler: handler)
+
+        await fulfillment(of: [handler.timerDidFire], timeout: 0.01)
+        XCTAssertEqual(handler.firedCount, 1)
+    }
+
+    func testCancel() async throws {
+        let loop = MultiThreadedEventLoopGroup.singleton.next()
+
+        let handler = MockTimerHandler()
+        handler.timerDidFire.isInverted = true
+
+        let handle = loop.scheduleCallback(in: .milliseconds(1), handler: handler)
+        handle.cancel()
+
+        await fulfillment(of: [handler.timerDidFire], timeout: 0.01)
+        XCTAssertEqual(handler.firedCount, 0)
+    }
+
+    func testShutdown() async throws {
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        let loop = group.next()
+        defer { try! group.syncShutdownGracefully() }
+
+        let handler = MockTimerHandler()
+        handler.timerDidFire.isInverted = true
+
+        _ = loop.scheduleCallback(in: .milliseconds(100), handler: handler)
+        try await group.shutdownGracefully()
+
+        await fulfillment(of: [handler.timerDidFire], timeout: 0.01)
+        XCTAssertEqual(handler.firedCount, 0)
+    }
+}
+
+fileprivate final class MockTimerHandler: NIOScheduledCallbackHandler {
+    var firedCount = 0
+    var timerDidFire = XCTestExpectation(description: "Timer fired")
```

Comment:
Maybe @FranzBusch remembers, but isn't there an issue with expectations and sendability? Should we just rely on a promise being completed instead?